### PR TITLE
ksw/2326 using shifted frequency data for filtering

### LINKED ISSUE #2326: [spectral line query] incorrect frequency filtering in the shifted frequency column
**Describe the bug**
With v4.0, when we apply a frequency range filter in the "shifted frequency" column, apparently the filter is applied to the "rest frequency" column instead. Therefore it is hard to select transitions based on the filtered "shifted frequency" column. 

**To Reproduce**
1. search for transitions between 84000 MHz and 85000 MHz with an intensity limit -6.
2. apply a velocity shift of 98 km/s and then we see the shifted frequency column is updated accordingly.
3. apply a filter as `84545..84547` in the shifted frequency column and we will see 4 transitions left.
4. observe the values in the shifted frequency column and in the rest frequency column to see the error. 

Screenshot 2024-01-03 at 13 50 39

**Expected behavior**
When we set a frequency range as a filter to the "shifted frequency" column, we should use the shifted values as the input for frequency filtering.

**Platform info (please complete the following information):**
 - OS [e.g. macOS Monterey]: macOS Ventura
 - Browser [e.g. chrome, safari, electron app]: v4-stable electron app
 - Browser version [e.g. 22]: 
 - Backend branch [e.g. dev, v3b2 release]: dev, v4-stable
 - Frontend branch [e.g. dev, v3b2 release]: dev, v4-stable

### PR DESCRIPTION
**Pull request by kswang1029**
**Description**
This is to close #2326. The root cause is when we do `filter`, the data used for filtering the "shifted frequency" column is always with v=0 (ie no doppler shift). Therefore we see unexpected results as described in #2326. This simple (but may not be ideal) solution is to apply doppler shift to the data and use it instead for filtering. 

Tested by applying filters to the shifted frequency column, rest frequency column, and any combination of the two as well as other string/number columns. The results are fine as far as I can tell.

**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] changelog updated / ~no changelog update needed~
- [x] ~unit test added~ (for functions with no dependenies)
- [x] ~API documentation added~ (for public variables and methods in stores)

For dependencies:
- [x] e2e test passing / corresponding fix added / ~new e2e test created~
- [x] ~protobuf version bumped~ / no protobuf version bumped needed
- [x] ~protobuf updated to the latest dev commit~ / no protobuf update needed
- [x] ~corresponding ICD test fix added (`BackendService` changed)~ / no ICD test fix needed (`BackendService` unchanged)
- [x] ~user manual prepared~ (for large new features)

### CODE REVIEW

File: src/stores/Widgets/SpectralLineQueryWidgetStore/SpectralLineQueryWidgetStore.ts [hunk 1]
```diff
@@ -317,7 +317,12 @@ export class SpectralLineQueryWidgetStore {
             if (filterString !== "") {
                 const column = this.queryResult.get(controlHeader.columnIndex);
```
Comment (loveluthien) [inline src/stores/Widgets/SpectralLineQueryWidgetStore/SpectralLineQueryWidgetStore.ts:318]: How about just replacing ```queryResult``` with ```filterResult```? My quick test works, but I am not sure if there are other effects.

Comment (loveluthien) [reply src/stores/Widgets/SpectralLineQueryWidgetStore/SpectralLineQueryWidgetStore.ts:318]: I saw the effect. Forget about this.

**Review by loveluthien — APPROVED** [review]
It looks good.

**Comment by YuHsuan-Hwang** [comment]
If we apply filtering to shifted freq first and then apply a new redshift, do we need to update the filtering result?

File: src/stores/Widgets/SpectralLineQueryWidgetStore/SpectralLineQueryWidgetStore.ts [hunk 2]
```diff
@@ -317,7 +317,12 @@ export class SpectralLineQueryWidgetStore {
...
                 const dataType = column.dataType;
-                const data = column.data;
+                const data =
+                    controlHeader.columnIndex !== SHIFTIED_FREQUENCY_COLUMN_INDEX
+                        ? column.data
+                        : column.data.map(value => {
+                              return isFinite(value) ? value * this.redshiftFactor : undefined;
```
Comment (crocka) [inline src/stores/Widgets/SpectralLineQueryWidgetStore/SpectralLineQueryWidgetStore.ts]: In JS, we can just write ```column.data.map(value => isFinite(value) ? value * this.redshiftFactor : undefined)```.

Comment (kswang1029) [reply src/stores/Widgets/SpectralLineQueryWidgetStore/SpectralLineQueryWidgetStore.ts]: thanks. revised.

**Comment by kswang1029** [comment]
> If we apply filtering to shifted freq first and then apply a new redshift, do we need to update the filtering result?

Good point. I tried to address this with [e9cfb15](https://github.com/CARTAvis/carta-frontend/pull/2328/commits/e9cfb1502c5548f581427c86f26365d4ab9a7032). Please have a look.

**Review by loveluthien — APPROVED** [review]
It is good no need to click apply filter again when changing the velocity.

**Review by YuHsuan-Hwang — APPROVED** [review]
The fix looks good. 👍

**Review by crocka — APPROVED** [review]

**Pull request merged**